bika/lims/validators.py

```diff
@@ -77,73 +74,119 @@ def __call__(self, value, *args, **kwargs):
             field_index = accessor.__name__
 
-        # 3. Check if the field index is in the indexes
-        # Field is indexed, use the catalog instead of objectValues
-        parent_path = api.get_parent_path(instance)
-        portal_type = instance.portal_type
-        catalog_query = {"portal_type": portal_type,
-                         "path": {"query": parent_path, "depth": 1}}
-
-        # We try here to avoid waking up all the objects, because this can be
-        # likely very expensive if the parent object contains many objects
-        if fieldname in catalog.indexes():
-            # We use the fieldname as index to reduce the results list
-            catalog_query[fieldname] = to_utf8(safe_unicode(value))
-            parent_objects = map(api.get_object, catalog(catalog_query))
-        elif field_index and field_index in catalog.indexes():
-            # We use the field index to reduce the results list
-            catalog_query[field_index] = to_utf8(safe_unicode(value))
-            parent_objects = map(api.get_object, catalog(catalog_query))
-        else:
-            # fall back to the objectValues :(
-            parent_object = api.get_parent(instance)
-            parent_objects = parent_object.objectValues()
+        # return if the field is not indexed
+        if field_index not in catalog.indexes():
+            return None
+
+        # build a catalog query
+        query = {
+            "portal_type": api.get_portal_type(context),
+            "path": {
+                "query": api.get_parent_path(context),
+                "depth": 1,
+            }
+        }
+        query[field_index] = value
+        logger.info("UniqueFieldValidator:Query={}".format(query))
+        return query
+
+    def __call__(self, value, *args, **kwargs):
+        context = kwargs['instance']
+        uid = api.get_uid(context)
+        field = kwargs['field']
+        fieldname = field.getName()
+        translate = getToolByName(context, 'translation_service').translate
+
+        # return directly if nothing changed
+        if value == field.get(context):
+            return True
+
+        # Fetch the parent object candidates by catalog or by objectValues
+        #
+        # N.B. We want to use the catalog to speed things up, because using
+        # `parent.objectValues` is very expensive if the parent object contains
+        # many items and causes the UI to block too long
+        catalog_query = self.make_catalog_query(context, field, value)
+        parent_objects = self.query_parent_objects(
+            context, query=catalog_query)
 
         for item in parent_objects:
-            if hasattr(item, 'UID') and item.UID() != instance.UID() and \
+            if hasattr(item, 'UID') and item.UID() != uid and \
                fieldname in item.Schema() and \
-               str(item.Schema()[fieldname].get(item)) == str(value):
+               str(item.Schema()[fieldname].get(item)) == str(value).strip():
```

Comment:
Yes... unfortunately, I am afraid we have some accessors that may return a value not consistent with the value stored in the Schema's field... Don't think there are too many places, but indeed, this is something we need to find some time to look in detail. Maybe a function in the api like `get_value(obj, field_name, default=None)` to deal with this would be helpful on this regard.